### ISSUE TITLE
Form still dirty after `initialize(formName, newData, true)`

### ISSUE THREAD

**Issue by ncphillips** [erikras/redux-form#1902]
We're using redux form to edit some pages, and I'm having some trouble re-initializing the form after saving.

The submit callback is 

```
const mapDispatchToProps = (dispatch) => ({
 submit: (page) => dispatch(updatePage(page)).then((page) => {
      dispatch(initialize(`page/${page.slug}`, page, false));
      dispatch(reset(`page/${page.slug}`));
      return page;
    });
});
```

This seems to be working as expected. The current and initial values for the form are equal, however the `dirty` flag is set to 'true' until I reload the page.

Any ideas on why this might be happening?

This seems to be effecting the state as I expected, the form values stay the same, but the

**Comment by ncphillips** [comment]
I also have a button to manually reset the form, and I just noticed that after this runs that button no longer effects the form.

**Comment by ncphillips** [comment]
I've written a test for this hopefully @erikras or someone can make it pass, because I can't. 

Here's the test, including some console.logs that I've put in.

**reduxForm.spec.js**
```jsx
    it.only('should reset to latest pristine state when reset() called, () ', () => {
      let renderCount = 0
      const store = makeStore({})
      const input = createSpy(props => (
         console.log(`RENDERING ${renderCount++} : `, props.input.value) 
         || <input {...props.input}/>).andCallThrough()
      )

      const Form = () => (
        <form>
          <Field name="bar" component={input} type="text"/>
        </form>
      )

      const Decorated = reduxForm({
        form: 'testForm',
        initialValues: { bar: 'initialBar' }
      })(Form)

      const dom = TestUtils.renderIntoDocument(
        <Provider store={store}>
          <Decorated/>
        </Provider>
      )
      
      const stub = TestUtils.findRenderedComponentWithType(dom, Decorated)

      inputOnNthRender(input, 0).onChange('newBar')

      console.log("Initialize");
      store.dispatch(initialize('test', { bar : 'newBar' }, true));

      console.log("Reset");
      stub.reset()

      expect(inputOnNthRender(input, 2).value).toBe('newBar')
    })
    
    const inputOnNthRender = (component, n) => component.calls[ n ].arguments[ 0 ].input
```

I also added a console.log to the reducer

**reducer.js**
```javascript
[INITIALIZE](state, { payload, meta: { keepDirty } }) {
      // Do stuff here...
      result = setIn(result, 'values', newValues)
      result = setIn(result, 'initial', mapData)
      console.log('Result after Initialize', result);
      return result
    },
```

**OUTPUT**
```
Result after Initialize: { values: { bar: 'initialBar' }, initial: { bar: 'initialBar' } }
FORM RENDER 0 Initial Values :    { bar: 'initialBar' }
INPUT RENDER 0 Value :  initialBar

Change Value
INPUT RENDER 1 Value :  newBar
FORM RENDER 1 Initial Values :  { bar: 'initialBar' }

Reinitialize
Result after Initialize : { values: { bar: 'newBar' }, initial: { bar: 'newBar' } }

Reset
INPUT RENDER 2 Value : initialBar
FORM RENDER 2 Initial Values :  { bar: 'initialBar' }
```

**Comments**

For some reason, the call to `initialize` updates the store, but does not make the Form or it's Field's update.

**Comment by ncphillips** [comment]
The issue is definitely in the mapStateToProps line 439 in reduxForm.js 

```javascript
(state, props) => {
  // ....
  const initial = initialValues || stateInitial || empty
  // ....
}
```

**Comment by ncphillips** [comment]
Further investigation shows this line of thought is at least partially incorrect. 

I logged out `stateInitial` and it turns out the new state `{bar: 'newBar'}` never actually gets passed to the ReduxForm.

Digging further, I found `byForm` and added a log on INITIALIZE

```javascript
 const byForm = (reducer) =>
    (state = empty, action = {}) => {
      const form = action && action.meta && action.meta.form
      if (!form) {
        return state
      }
      if (action.type === DESTROY) {
        return deleteInWithCleanUp(state, action.meta.form)
      }
      const formState = getIn(state, form)
      const result = reducer(formState, action)
      if (action.type === INITIALIZE) {
        console.log('INITIALIZED', result, setIn(state, form, result));
      }
      return result === formState ? state : setIn(state, form, result)
    }
```

This logger shows

```javascript
// INITIALIZED 
let result = { values: { bar: 'newBar' }, initial: { bar: 'newBar' } } 

let setInResult = { 
  testForm: { 
    values: { bar: 'newBar' },
    initial: { bar: 'initialBar' },
    registeredFields: [ [Object] ] },
  test: { values: { bar: 'newBar' }, initial: { bar: 'newBar' } } 
}
```

For some reason, `setIn` is removing the word `Form` from the form name!

**Comment by ncphillips** [comment]
I'm an idiot

**Comment by ncphillips** [comment]
Here's a test describing the expected behaviour
```jsx
it.only('should not be dirty after initialize()', () => {
      const store = makeStore({})
      const input = createSpy(props => <input {...props.input}/>).andCallThrough()

      const Form = createSpy((props) => (
        <form>
          <Field name="bar" component={input} type="text"/>
        </form>
      )).andCallThrough()

      const Decorated = reduxForm({
        form: 'testForm',
        initialValues: { bar: 'initialBar' }
      })(Form)

      const dom = TestUtils.renderIntoDocument(
        <Provider store={store}>
          <Decorated/>
        </Provider>
      )

      const stub = TestUtils.findRenderedComponentWithType(dom, Decorated)

      inputOnNthRender(input, 0).onChange('newBar')

      expect(propsOnNthRender(Form, 1).dirty).toBe(true)

      store.dispatch(initialize('testForm', { bar : 'newBar' }))
      
      expect(propsOnNthRender(Form, 1).dirty).toBe(false)
    })
```

**Comment by skolmer** [comment]
@ncphillips let's create a fork containing this test. I will help to work on a solution later this week. will add a test to tackle the reset issue I have reported in #1776

**Comment by ncphillips** [comment]
Thanks @skolmer, I actually figured out how to make it work, but it broke a couple tests that were relying on render counting to validate their behaviour. I'll fork the project and submit a PR here shortly.

**Comment by erikras** [comment]
Fix published in [`v6.2.0`](https://github.com/erikras/redux-form/releases/tag/v6.2.0).
